SSR user authentication

sdelfi opened the pull request:
Добавил авторизацию в SSR и автоматический проброс preloadedState для store. 

Т.е. сервер при рендеринге проверяет cookie из запроса и пробрасывает их в axios запросы, которые делаются в клиентском пакете, но на стороне сервера.

Т.е. теперь рендерятся корректно роуты, которые закрыты под авторизацией (без редиректов). 

Переделал leaderboard для предзагрузки данных на сервере + подправил тесты и добавил новый, проверяющий, что таблица рендерится. 

С leaderboard только не смог разобраться - SSR ждет, пока данные с сервера Яндекс будут получены и после рендерит таблицу (по аналогии с Root лоадером). 
Но почему-то рендерится fallback loader в месте, где должна быть таблица:
![image](https://user-images.githubusercontent.com/1927032/215347824-c66d1027-8fc9-407d-807d-07f2bdf13bc4.png)

А сама таблица с данными, которая должна отрендериться на месте этого loader отображается ниже и у нее атрибут hidden
![image](https://user-images.githubusercontent.com/1927032/215347866-b5ebc12a-f298-4fc9-b941-14694fbe6fee.png)

Т.е. данные подгрузились корректно, но React почему-то не поставил на место таблицу с данными.

sdelfi committed:
feat(ssr): add user authentication on server

add SSR store state preloading
update leaderboard structure (preload leaderboard data)
update leaderboard tests

Olegas commented on packages/client/ssr.tsx:
```diff
@@ -8,6 +9,11 @@ import { routes } from './src/config/router';
+  /** Пробрасываем Cookie в серверный Axios для получения данных из Яндекс API, закрытых авторизацией */
+  if (request.headers.cookie) {
+    axios.defaults.headers.common['Cookie'] = request.headers.cookie;
+  }
```
Опасно. Представь что тебе одновременно пришло два запроса на рендер страницы от разных пользователей. И вот у тебя уже для первого пользователя полетели данные от второго

sdelfi replied:
А как быть?
Как я понимаю, этот код выполнится для конкретного запроса. Разве нет?

Мне нужно для запросов на сервере, которые делаются в пакете client отправлялись авторизационные куки, как это происходит от клиента.

В данном случае render из ssr.ts вызывается сервером в момент запроса от пользователя. Т.е. берутся куки из запроса пользователя, что пришли на сервер и пробрасываются в клиентскую часть, где делаются запросы в режиме реального времени именно для этого запроса.
Т.е. запрос от другого пользователя пойдет по такой же цепочке и запросы будут именно с его куками идти.
Это как я понял принцип работы.

Olegas replied:
> Как я понимаю, этот код выполнится для конкретного запроса. Разве нет?

Этот код - для конкретного. Но вот axios ты правишь глобальный.

Что можно попробовать.
1. Используя axios.create создать экземпляр axios с конкретными настройками
2. Обернуть приложение контекстом, в который положить этот экземпляр
3. В точках, где нужен axios, брать не глобальный, а вот этот вот из контекста. 

Кажется это можно сделать. Ведь загрузка у тебя в loader'ах а их ты передаешь в параметры роута уже внутри приложения. Можно в этой точке добраться до axios из контекста и на основе него сделать loader

Как-то так
```  
 const leaderboardLoader = useLeaderboardLoader()

 function useLeaderboardLoader() {
       const localAxios = useLocalAxios(); // тут достаем его из контекста
       return useMemo(() => {
           return function loader() { 
              // load smth with localAxios
           }
       }, [localAxios]);
}
```

Olegas replied:
https://axios-http.com/docs/instance

sdelfi replied:
Спасибо за пояснение, попробую разобраться.
Пока все ещё не понимаю, как куки одного пользователя попадет в запрос другому, ведь глобальные настройки axios устанавливаются в рамках текущего запроса. Завтра на свежую голову ещё посмотрю.
Но тема интересная

Olegas replied:
> ведь глобальные настройки axios устанавливаются в рамках текущего запроса

Вот здесь ошибка. Это не верное утверждение.

Смотри, node выполняет все запросы в рамках одного процесса. Ты можешь это проверить сделав глобальную переменную и меняя ее на каждый запрос и возвращая пользователю. 

когда ты пишешь `import axios from 'axios'` ты создаешь глобальную переменную. Когда мы меняешь axios.defaults ты меняешь их все так же в рамках глобальной переменной.

sdelfi replied:
Теперь понятно, интересно.

sdelfi replied:
@Olegas спасибо за разъяснение. Провел эксперимент и проблема воспроизвелась, просто открыв в двух браузерах сайт. 
Очень полезное замечание и "тонкость", которая могла бы привести к серьезным проблемам. 

Я залил исправления. Я не стал использовать контекст. У нас есть httpClient файл, где был настроен interceptor для изменения формата ответа ошибок от Яндекс API. А далее мы импортировали axios из этого файла.

Там же сделал инстанс axios и сделал проброс авторизационных cookie в defaults этого инстанса. Проверил - все работает корректно. 
У нас импорт идет только один раз и в данном случае создается только один инстанс, даже при повторном импорте, это дополнительно проверил.

sdelfi replied:
Плюс если использовать контекст, то у меня получается проблема с очередностью.

Мне нужно ставить заголовки вначале render. Если я делал это ниже по коду, например, то не работало. Полагаю, в этой магии, что здесь - что-то происходит и уже после этого httpClient.defaults.header не задать. Вернее, задать можно, но в запросы не попадают дефолтные заголовки:
![image](https://user-images.githubusercontent.com/1927032/215546294-60c2b3b0-dfe4-4b67-ad11-a3626914617c.png)

А контекст провайдер оборачивает уже ниже приложение. В том числе
![image](https://user-images.githubusercontent.com/1927032/215545775-f4e01715-eea3-43ed-ac30-847ea09ad17c.png)


В целом, это решаемо.

Но дальше есть 2 момента: 
В rootLoader нельзя использовать хуки, в т.ч. и контекст, т.е. туда не получить инстанс клиента. 
Также у нас есть обертка HTTP.ts, внутри которой тоже импортируется клиент. Так вот там useContext, логично, тоже не заиспользовать.

Т.е. либо у нас архитектурно как-то неверно сделано и переделывать много кода (при этом не ясно как в loader прокинуть инстанс клиента, либо использовать сделанный мною вариант (если, конечно, в нем нет каких-то критичных моментов, которые мне неизвестны)

Olegas replied:
> Там же сделал инстанс axios и сделал проброс авторизационных cookie в defaults этого инстанса. Проверил - все работает корректно.

А чем это отличается от того что было? Был один глобальный инстанс, стал другой такой же глобальный, пусть и со своими интерцепторами

sdelfi replied:
В таком виде авторизационные куки не делились между разными клиентами. 
Тогда не понимаю как можно еще реализовать, если есть в текущей реализации проблемы.
Описал выше о проблемах реализации через контекст, где доступ к инстансу внутри контекста из route loader не получить, т.к. там нельзя использовать хуки

Olegas commented on packages/client/ssr.tsx:
```diff
@@ -8,6 +9,11 @@ import { routes } from './src/config/router';
+
   const { query } = createStaticHandler(routes);
   const remixRequest = createFetchRequest(request);
   const context = await query(remixRequest);
```
А что это за магия? Расскажи плз? Тут какие-то части от фреймворка Remix?

sdelfi replied:
Это взял с примеров React router, чтобы завелся createStaticRouter,  без которого не работал корректно SSR, поэтому, увы, не подскажу что за магия, сам не до конца разобрался, а в документации у них этого ещё не написано, т.к. createStaticRouter только недавно внедрили.

sdelfi replied:
https://github.com/remix-run/react-router/blob/main/examples/ssr-data-router/src/entry.server.tsx

Olegas replied:
Кстати, я вижу что вы создаете context и передаете его в роутер.
А вот что я вижу в тайпингах аргументов, которые приходят в loader

```
interface DataFunctionArgs {
  request: Request;
  params: Params;
  context?: any;
}
```

А нет ли там вашего исходного запроса с нужными хэдерами или этого самого контекста? Кажется это тот самый вариант добросить кукисы до точки вызова загрузки данных

sdelfi replied:
@Olegas 

В request действительно есть исходные куки. Спасибо за наводку.

Тогда возникает другой вопрос. 

У нас часть запросов есть в лоадере, часть вне его.
Т.е. получается, что в лоадере мы во все запросы прокидываем из request.header.cookie авторизационную куку. 
А в остальное через контекст? 

Вариант 
authAPI.me(cookie) мне не нравится, т.к. делать для каждого api запроса параметр cookie, затем его пробрасывать так
![image](https://user-images.githubusercontent.com/1927032/216781591-7bb4e1cc-50f4-4956-a907-4e8e6cbf8e10.png)

Где дальше идет проброс запроса через HTTP._send -> httpClient - это выглядит костыльно как-то.

Была еще мысль сделать базовый класс BaseApi и в нем делать addCookie метод, от него ветвить AuthApi и другие, чтобы не дублировать код.
Но все равно получается, что в каждом методе AuthApi и других учесть наличие static cookie параметра, добавленного через addCookie - тоже выглядит как-то неправильно.

При этом, классы статичные, т.к. в контекст их инстансы не сохранить, т.к. нужен доступ из loader и тут встает вопрос - если статичный класс, мы в него добавили авторизационную куку - не будет ли такой же проблемы, как и с дефолтными настройками axios, где кука шарится между разными запросами?

..................

В общем, ряд экспериментов показал, что проблема осталась после последних правок, Вы правы.

Решил взглянуть на проблему иначе, отойдя от старых попыток.
Попробовал еще сделать класс с HTTP клиентом (где инициализируется инстанс axios) с использованием singleton, но тоже не работает.
Заметил, что первая инициализация - создание инстанса идет при запуске сервера, а дальше уже берется этот же инстанс. Откатывать не стал, хоть и не решило проблему.

В общем, проблему решил несколько проще. Я делал проверку, что если кука с запроса есть, то пишем в дефолт её. Убрал проверку. Т.е. теперь пишем в дефолт всегда, даже если в запросе куки пустые. Т.е. можно сказать, что мы в дефолт ставим всегда именну ту куку, которая прилитела в этом запросе. Проверил - все работает.

По факту. со стороны сервера, это единственная точка входа, где после используется axios, т.е. в других местах эта дефолтная кука не всплывает, а тут всегда записываются актуальные куки текущего запроса.

Olegas replied:
Я придумал. Вам не нужен HTTPClient, getInstance и т.п.
Общая идея - добавить на стороне сервера в глобальный (да) axios новый интерцептор, который будет перед запросом добавлять куки.
Где он их будет брать? 
В Node есть возможность сделать "переменную", которая будет локальной по отношению к текущей цепочке асинхронных процессов (называется TLS, Thread Local Storage)
В вашем случае такая цепочка будет = запросу и всем идущим далее асинхронным процессам.
Т.е. на старте запроса вы создаете такое хранилище, складываете в него кукисы и идете дальше. Axios внутри интерцентора (который объявлен только на сервере и просто отсутствует на клиенте) достает их этого хранилища данные о кукисах и подставляет в запрос.
Материалы по теме:
- https://nodejs.org/dist/latest-v18.x/docs/api/async_context.html#class-asynclocalstorage
- https://pjatk.in/tls-in-node.html

sdelfi replied:
Звучит интересно, ни разу не слышал. 
Спасибо, буду пробовать такой вариант. Получается, что клиент вообще ничего не будет знать о куки, получается, об этом позаботится сервер, как я понял.

Olegas replied:
Да, именно! Прелесть этого решения что оно полностью локализовано на сервере и не потребует правок в клиентской части вообще.

Я вчера пол проекта вашего нарефакторил с целью проброса кукисов до транспорта, и тут до меня дошло как вообще на изи все сделать )

sdelfi replied:
Отлично, спасибо еще раз за вовлеченность :)

Главное, чтобы Thread Local Storage хранил именно для цепочки запросов клиента. Смущает, что в названии Thread, ведь один thread обрабатывает как раз запросы от разных клиентов и шарит между ними данные. 
Но для начала изучу матчасть и попробую реализовать, выводы напишу после.

Olegas replied:
Да, ты абсолютно прав, Thread тут немного смущает. Но по факту все работает ) Thread Local это "калька" аналогичного поняти из других стэков, но суть в том что позволяет держать переменные в рамках обработки одного запроса.

sdelfi replied:
Тогда отлично. В выходные займусь этим, руки чешутся привести в порядок и влить этот PR в develop :)

DmitriiNazimov committed:
feat: add themization (#83)

* feat: add light theme

* feat: add ThemeToggle component

* feat: add themization API

* feat: add auth check middleware

* test(ThemeToggle): add tests

* refactor(checkAuthMiddleware): improve security

* refactor: implement review recomendation

sdelfi committed:
Merge branch 'develop' into task/add-ssr-authorization

sdelfi committed:
fix(security): fix security issue related to sharing of authentication cookies between requests

Olegas commented:
> Т.е. данные подгрузились корректно, но React почему-то не поставил на место таблицу с данными.

Это "последствия" того, что используется Suspense и renderToPipeableStream
Штука в том что React доставляет приложение кусками. Если приглядеться то видно что рядом идет скрипт, который один кусок заменяет на другой. Мы просто как будто сразу видим результат, по факту он прилетает к нам частями.

Что бы все прилетело разом, надо использовать onAllReady вместо onShellReady. Если поменять в текущей реализации, можно будет увидеть как в результирующем HTML пришла сразу таблица лидерборда например. Правда, без данных. Данные будут только в initialState

sdelfi commented:
Спасибо, с onAllReady и onShellReady поэкспериментирую.

sapomaro committed:
fix: game bugs (#84)

Фиксы:
- улучшена физика попадания снарядов в танки (исправлены баги с подрывом танков, когда снаряд не до конца попал в цель или когда попал, но танк не взорвался);
- подсчёт времени для статистики/лидерборда теперь останавливается при паузе;
- враги теперь появляются не все сразу, а постепенно;
- добавлены попытки респауна игрока, если на соответствующей начальной позиции стоит вражеский танк;
- изменены байдинги клавиш, чтобы избежать случайных нажатий не туда при активной игре;
- убран баг, когда перед началом анимации спауна танка (звезды) появлялось его изображение;
- исправлен баг, когда анимация движущегося танка отображалась на других экранах вне главного меню;
- оптимизировано отображение экранов интро и паузы при одновременном появлении (когда игрок ставит игру на паузу в самом начале игры).

sdelfi commented:
> Что бы все прилетело разом, надо использовать onAllReady вместо onShellReady. Если поменять в текущей реализации, можно будет увидеть как в результирующем HTML пришла сразу таблица лидерборда например. Правда, без данных. Данные будут только в initialState

@Olegas да, так и есть, как и написали. Но теперь вылезает ошибка hydration. 
Не совсем тогда понятно - почему не происходит рендер данных в таком случае?

Olegas commented:
> > Что бы все прилетело разом, надо использовать onAllReady вместо onShellReady. Если поменять в текущей реализации, можно будет увидеть как в результирующем HTML пришла сразу таблица лидерборда например. Правда, без данных. Данные будут только в initialState
> 
> @Olegas да, так и есть, как и написали. Но теперь вылезает ошибка hydration. Не совсем тогда понятно - почему не происходит рендер данных в таком случае?

Я думаю тут проблема в том, что получилось замешано два фреймворка react-router и redux каждый из которых хочет сам управлять жизненным циклом отрисовки. Хоть в доке к react-router и сказано что они не ограничивают как грузить данные, но по факту получается что они не могут дождаться корректно загрузки

Но тут пока только мое ИМХО. Я вчера пару часов сидел с вашим проектом, пытался его по нормальному запустить - ничего лучше не вышло (

sdelfi commented:
Спасибо за попытку разобраться. С leaderboard, полагаю, пока можно пренебречь.
А вот с Axios еще провожу эксперименты.

sdelfi committed:
Merge branch 'develop' into task/add-ssr-authorization

Olegas commented:
> Спасибо за попытку разобраться. С leaderboard, полагаю, пока можно пренебречь. А вот с Axios еще провожу эксперименты.

"по нормальному" я не имел ввиду что плохо сделано, имел ввиду что хотелось сделать что бы все дожидалось и грузилось. Вообще у вас там титанический труд конечно, вы молодцы )

sdelfi committed:
fix: fix authorization security issue

sdelfi committed:
fix(tests): fix tests

sapomaro commented on packages/client/src/config/leaderboardRoute.tsx:
```diff
@@ -0,0 +1,26 @@
+/**
+ * Предзагрузка данных leaderboard
+ */
+export const leaderboardLoader: LoaderFunction = () => {
+  const leaderboardData = leaderboardAPI.getLeaderboard({
+    ratingFieldName: LEADERBOARD_SORT_FIELD,
+    cursor: LEADERBOARD_DEFAULT_PAGE,
+    limit: LEADERBOARD_RECORDS_DISPLAY_LIMIT,
+  });
+  return { leaderboardData };
+};
```
У нас такой же метод есть в `leaderboardThunks`. Насколько я понял, подружить `Redux` и `React Router` не получилось, поэтому нужно дублирование?

sdelfi replied:
Верно, redux и thunks не доступны в лоадере

sapomaro commented on packages/client/src/utils/HTTP/httpClient.ts:
```diff
@@ -3,16 +3,35 @@ import axios, { AxiosError } from 'axios';
+  }
+
+  private initInterceptors() {
+    // Check if error response body contains reason of error
+    this.httpClient.interceptors.response.use(
+      response => response,
+      (error: AxiosError<Record<string, unknown> | APIError>): Promise<AxiosError> => {
+        if (error instanceof AxiosError && error.response && apiHasError(error.response.data)) {
+          return Promise.reject(new Error(error.response.data.reason));
+        }
+
+        return Promise.reject(error);
+      }
+    );
   }
-);
+}
```
Столкнулся с такой вещью, которая к SSR отношения не имеет. Если заходить с клиентской сборки (порт 3000), то кидает следующую ошибку в консоли (темизация реализована только у нас на сервере, который не запущен):

```
XHRGEThttps://ya-praktikum.tech/api/v2/themization
[HTTP/2 404 Not Found 15ms]

Uncaught (in promise) Error: Not found
    initInterceptors http://localhost:3000/assets/js/app-a67191d2.js:100
    promise callback*request http://localhost:3000/assets/js/app-a67191d2.js:100
    em http://localhost:3000/assets/js/app-a67191d2.js:98
    _send http://localhost:3000/assets/js/app-a67191d2.js:100
    get http://localhost:3000/assets/js/app-a67191d2.js:100
    getUserTheme http://localhost:3000/assets/js/app-a67191d2.js:104
    rC http://localhost:3000/assets/js/app-a67191d2.js:104
    rC http://localhost:3000/assets/js/app-a67191d2.js:104
    ml http://localhost:3000/assets/js/app-a67191d2.js:32
    Kr http://localhost:3000/assets/js/app-a67191d2.js:32
    Kd http://localhost:3000/assets/js/app-a67191d2.js:32
    qn http://localhost:3000/assets/js/app-a67191d2.js:30
    py http://localhost:3000/assets/js/app-a67191d2.js:32
    er http://localhost:3000/assets/js/app-a67191d2.js:32
    nh http://localhost:3000/assets/js/app-a67191d2.js:32
    _ http://localhost:3000/assets/js/app-a67191d2.js:17
    ge http://localhost:3000/assets/js/app-a67191d2.js:17
    EventHandlerNonNull* http://localhost:3000/assets/js/app-a67191d2.js:17
    <anonymous> http://localhost:3000/assets/js/app-a67191d2.js:17
app-a67191d2.js:100:14058
```

Получается, что ошибки при неполучении данных через API на клиенте никак не обрабатываются. Предлагаю рассмотреть этот момент отдельно - вне текущей задачи.

sdelfi replied:
Явно не хватает обработчика ошибок при запросе темы. Возможно, то связано с интерсептором, который ловит и конвертирует ошибки от яндекс api

sapomaro reviewed (approved):
Потестил, с авторизацией работает. Лидерборд приходит с сервера со всеми данными в __PRELOADED_STATE__. Станислава не останавливают никакие трудности)

sdelfi committed:
fix(themization): add try/catch for get request

Olegas commented:
Я увидел новые комиты и комментарии. Ушел думать как помочь...

sdelfi committed:
Merge branch 'develop' into task/add-ssr-authorization

sdelfi committed:
refactor: move auth cookie interceptor to server

sdelfi commented:
@Olegas залил обновление, внедрил TLS. Спасибо за подсказку. 

Единственный момент смущает, правильно ли сделал: в HTTPClient у нас создается индивидуальный инстанс и к нему не цепляется interceptor, который задал на сервере. В итоге я прикрепил к клиенту request interceptor из глобального инстанса.  Если не сделать, то не работало.
Не хотелось просто на клиенте все-таки использовать глобальный инстанс клиента.

Также перенес SSR на сервере в отдельный файл, чтобы index.ts был почище. Давно хотел сделать это.

Olegas commented:
> Не хотелось просто на клиенте все-таки использовать глобальный инстанс клиента.

На клиенте нет ничего плохого в глобальном инстансе клиента.

Olegas reviewed (approved):
👍
